aten/src/ATen/core/Generator.h

```diff
@@ -49,45 +51,48 @@
 
 namespace at {
 
-// The default seed is selected to be a large number
-// with good distribution of 0s and 1s in bit representation
-constexpr uint64_t default_rng_seed_val = 67280421310721;
-
 struct CAFFE2_API Generator {
-  // Constructors
-  Generator(Device device_in, DispatchKeySet key_set);
+  Generator() {}
 
-  // Delete all copy and move assignment in favor of clone()
-  // method
-  Generator(const Generator& other) = delete;
-  Generator(Generator&& other) = delete;
-  Generator& operator=(const Generator& other) = delete;
+  explicit Generator(std::shared_ptr<c10::GeneratorImpl> gen_impl)
+   : impl_(std::move(gen_impl)) {
+    if (impl_.get() == nullptr) {
+      throw std::runtime_error("GeneratorImpl with nullptr is not supported");
+    }
+  }
 
-  virtual ~Generator() = default;
-  std::shared_ptr<Generator> clone() const;
+  // TODO(pbelevich): delete this after replace Generator generator = nullptr with c10::optional<at::Generator> = c10::nullopt
+  Generator(std::nullptr_t gen_impl) {}
 
-  // Common methods for all generators
-  virtual void set_current_seed(uint64_t seed) = 0;
-  virtual uint64_t current_seed() const = 0;
-  virtual uint64_t seed() = 0;
-  Device device() const;
+  bool operator==(const Generator& rhs) const {
+    return (!(this->impl_) && !(rhs.impl_)) || (this->impl_ == rhs.impl_);
```

Comment:
Can't this just be `this->impl_ == rhs->impl_`? Because if they're both nullptr they'll compare equal.

Reply:
addressed